common/src/main/java/org/conscrypt/Hkdf.java

```diff
@@ -0,0 +1,229 @@
+     * @throws InvalidKeyException if the salt is not suitable for use as an HMAC key
+     */
+
+    public byte[] extract(byte[] salt, byte[] ikm) throws InvalidKeyException {
+        Objects.requireNonNull(salt);
+        Objects.requireNonNull(ikm);
+        if (salt.length == 0) {
+            salt = new byte[getMacLength()];
+        }
+        Preconditions.checkArgument(ikm.length > 0, "Empty keying material");
+        Mac mac = getMac(salt);
+        return mac.doFinal(ikm);
+    }
+
+    /**
+     * Performs an HKDF expand operation as specified in RFC 5869.
+     *
+     * @param prk a pseudorandom key of at least HashLen octets, usually the output from the
+     *            extract step. Where HashLen is the key size of the underlying Mac
+     * @param info optional context and application specific information, can be zero length
+     * @param length length of output keying material in bytes (<= 255*HashLen)
+     * @return output of keying material of length bytes
+     * @throws InvalidKeyException if prk is not suitable for use as an HMAC key
+     * @throws IllegalArgumentException if length is out of the allowed range
+     */
+    public byte[] expand(byte[] prk, byte[] info, int length) throws InvalidKeyException {
+        Objects.requireNonNull(prk);
+        Objects.requireNonNull(info);
+        Preconditions.checkArgument(length >= 0, "Negative length");
+        Preconditions.checkArgument(length < 255 * getMacLength(), "Length too long");
+        Mac mac = getMac(prk);
+        int macLength = getMacLength();
+
+        // buffer[] is used for the data to be hashed each round
+        int bufferLength = info.length + 1;                // Initially t.length = 0
+        int maxBufferLength = macLength + info.length + 1; // Later, t.length = macLength
+        byte[] buffer = new byte[bufferLength];
+
+        byte[] t = new byte[0];
+        byte[] output = new byte[length];
+        int outputOffset = 0;
+        byte[] counter = new byte[] { 0x00 };
+        while (outputOffset < length) {
+            counter[0]++;
+            t = mac.doFinal(concat(buffer, t, info, counter));
+            int size = Math.min(macLength, length - outputOffset);
+            System.arraycopy(t, 0, output, outputOffset, size);
+            outputOffset += size;
+
+            // Re-allocate data buffer if needed (will only happen once, on iteration 2)
+            if (bufferLength != maxBufferLength) {
+                bufferLength = maxBufferLength;
+                buffer = new byte[bufferLength];
+            }
+        }
+        return output;
+    }
+
+    private byte[] concat(byte[] target, byte[]... sources) {
+        int offset = 0;
+        for (byte[] source : sources) {
+            System.arraycopy(source, 0, target, offset, source.length);
+            offset += source.length;
+        }
+        if (offset != target.length) {
+            throw new IllegalStateException("concat() error");
```

Comment:
Done.

I think it's fine to throw IOOBE, all args to this function are under the control of the class, so the only way it can start throwing IOOBE is during future changes.

In fact since I removed a lot of copying, this is only called from one site and so it's tempting to inline in.  Leaving it this way for now for clarity.